pkg/kubelet/status/status_manager.go

```diff
@@ -898,12 +900,12 @@ func (m *manager) canBeDeleted(pod *v1.Pod, status v1.PodStatus, podIsFinished b
 		return false
 	}
 	// Delay deletion of pods until the phase is terminal.
-	if !podutil.IsPodPhaseTerminal(pod.Status.Phase) {
+	if !podutil.IsPodPhaseTerminal(status.Phase) {
```

Comment:
Can you explain the rationale behind this change? Why should we look at `status.Phase` vs `pod.Status.Phase` and why are the different in the first place here? I think that's at the root of the issue here.

Reply:
> Can you explain the rationale behind this change? Why should we look at status.Phase vs pod.Status.Phase and why are they different in the first place here?

To restore the answer from `needsUpdate` (to return true as before this PR). Returning `pod.Status.Phase` or `status.Phase` does not matter when the function is called from `syncPod`, because the pod is already updated at this point: https://github.com/kubernetes/kubernetes/blob/8b7c7d783908efb53056fcae121ed18e117a8f2e/pkg/kubelet/status/status_manager.go#L849. I have missed the fact it makes a difference for the other path. In the other case the `pod.Status.Phase` returns `Pending` for a non-admissible pod created while the kubelet was down (as in the e2e test)

>  I think that's at the root of the issue here.

Not the entire issue. The second part of the fix is important, because `podIsFinished` returns false.

Reply:
> I have missed the fact it makes a difference for the other path

What do you mean here by the other path? It seems to be that one of the surprises here is that within `needsUpdate` the pod (as fetched from pod manager) and status as passed to the function have different phase (i.e. the pod one is pending, the status is Failed).

Regarding the second part of the fix, why do we also need to consult the pod worker? Is it because `podIsFinished` is never true because no-one ever called `TerminatePod`?

Reply:
> What do you mean here by the other path? It seems to be that one of the surprises here is that within needsUpdate the pod (as fetched from pod manager) and status as passed to the function have different phase (i.e. the pod one is pending, the status is Failed).

I think it would also have a different phase on 1.26, can confirm. It is only in the `syncPod` that the `pod` object is updated based on the status after successful PATCH.

> Regarding the second part of the fix, why do we also need to consult the pod worker with m.podDeletionSafety.PodIsFinished? Is it because podIsFinished is never true because no-one ever called TerminatePod?

Yes, but `TerminatePod` wasn't called in 1.26, can double check.

Reply:
Pod worker is authoritative for what the status of the pod is "if the kubelet was the one who started it".  After a restart, the apiserver / runtime are sources of truth for what the status of the pod is.  So the phase on the pod object passed to pod worker isn't really "truth".

We should be calling terminatepod on pods that can never run now but may have run before (i.e. killing the pod).  Only if we are 100% certain the pod could never have been run should we bypass the pod worker (and even then it is problematic).

Reply:
That makes sense, I think one of the root issues in https://github.com/kubernetes/kubernetes/issues/118472 is that the pod was never admitted in the first place and is rejected by kubelet so it never ran. However, we never called `TerminatePod` on it, so it is blocking status manager to delete it.

I'm trying out an alternative approach of simply calling `TerminatePod` for pods rejected at kubelet admission time in https://github.com/kubernetes/kubernetes/pull/118599. Would be great @smarterclayton and @mimowo  feedback on that approach.

Reply:
I was thinking about an alternative approach where we call `TerminatePod` if the pod is rejected at kubelet admission time.  I have a WIP here - https://github.com/kubernetes/kubernetes/pull/118599, would be great to both of your feedback.

Reply:
Pushed a commit to eliminate the spurious check.

Reply:
Marking as resolved as this is outdated code change, as the current approach (of using HandlePodCleanups) does not modify this.